python/ql/lib/semmle/python/ApiGraphs.qll

```diff
@@ -591,8 +591,18 @@ module API {
         or
         // Subclassing a node
         lbl = Label::subclass() and
-        exists(DataFlow::Node superclass | pred.flowsTo(superclass) |
-          ref.asExpr().(PY::ClassExpr).getABase() = superclass.asExpr()
+        exists(PY::ClassExpr clsExpr, DataFlow::Node superclass | pred.flowsTo(superclass) |
+          clsExpr.getABase() = superclass.asExpr() and
+          // Potentially a class decorator could do anything, but we assume they are
+          // "benign" and let subclasses edges flow through anyway.
+          // see example in https://github.com/django/django/blob/c2250cfb80e27cdf8d098428824da2800a18cadf/tests/auth_tests/test_views.py#L40-L46
+          (
+            not exists(clsExpr.getADecorator()) and
+            ref.asExpr() = clsExpr
+            or
+            ref.asExpr() = clsExpr.getADecoratorCall() and
+            not exists(PY::Call otherDecorator | otherDecorator.getArg(0) = ref.asExpr())
+          )
```

Comment:
I am not saying we should do this, but it seems (from your test) that the code could be simplified as follows
```suggestion
          (
            ref.asExpr() = clsExpr
            or
            ref.asExpr() = clsExpr.getADecoratorCall()
          )
```
but possibly the code in the PR is more performant in that it creates fewer tuples?

Reply:
Yeah, I was going to point this out as well. In some sense, it might be more correct to treat the result of each decorator as a subclass (if, say, you want to use API graphs to figure out which subclasses are decorated with a particular decorator).

In terms of performance I wouldn't expect it to make much of a difference. Yes, it's more API graph nodes, but all of the "inner" decorators don't flow anywhere except into the decorator immediately preceding them.

Reply:
I did this to optimize for performance yes. I can see the argument that it's nice to be able to also be able to track this into the arguments of the decorators, so I'll accept the suggestion :+1: